Design/indexing_rollout_notes.md

```diff
@@ -0,0 +1,209 @@
+
+An initial survey shows that almost half of the occurrences are for workspace initialization.
+Those cases are usually comparatively easy and straightforward to replace by the new style of workspace factory functions defined in `DataObjects/WorkspaceCreation.h` that use `IndexInfo` to create a workspace.
+
+Other significant fractions of the occurrences are made up by logging and exceptions and by tests that initialize individual spectra.
+Furthermore there are comparisons between the respective spectrum and detector information in two workspaces.
+
+
+## General goals
+
+- Quite a few algorithms take an input workspace, extract information about spectrum numbers and detector IDs, and initialize a new workspace based on those data. Example include `SumSpectra`, `DiffractionFocussing`, and `ExtractSpectra`.
+  One goal of the indexing library is to remove explicit uses of spectrum numbers and detector IDs in as much code as possible.
+  Since we need to refactor algorithms similar to those mentioned above, it is probably a good idea to do it properly right away.
+  That is, we should *not* just refactor the manipulation of spectrum numbers and detector IDs via `IndexInfo`.
+  Instead, we should provide functionality for grouping and extraction of spectra to create a new `IndexInfo`, without the need for explicit handling of spectrum numbers and detector IDs.
+
+  As an example, consider
+
+  ```cpp
+  namespace Indexing {
+  IndexInfo extract(const IndexInfo &, const SpectrumIndexSet &);
+  IndexInfo group(const IndexInfo &, const std::vector<SpectrumIndexSet> &);
+  }
+  ```
+
+- Should we set default detector IDs?
+  This is was workspaces are doing currently, but it does not really make sense.
+  Probably we should find a better way if setting sensible detector IDs once an instrument has been set.
+  This is something that potentially needs to be decided together with other decisions regarding Instrument-2.0.
+
+
+## NearestNeighbours
+
+- Uses `specnum_t` as map key.
+- Should simply use spectrum index?
+- Would that make trouble with MPI? Need to have a detailed look.
+- See issue [#17372](https://github.com/mantidproject/mantid/issues/17372).
+
+
+## Logging and exception messages
+
+- Frequently spectrum numbers are included in log messages or exceptions.
+
+- Probably the easiest (overall) solution is to add means of getting this information automatically, e.g., with a helper function
+
+  ```cpp
+  namespace Logging {
+  std::string spectrumNumber(const MatrixWorkspace &workspace, size_t index);
+  }
+  ```
+
+## File- and UI output
+
+Similar to logging, there are also several algorithms that write, e.g., detector IDs to a file.
+Maybe a similar or even the same helper function as suggested above in the logging section could be used.
+
+
+## Obtaining detector information
+
+Frequently, detector IDs are used to obtain detector pointers or other detector-related information from the instrument.
+The IDs are not actually required here.
```

Comment:
True in the case that the client is not specifying the detector id. Instrument 2.0 essentially does not use this information since most operations are applied in bulk to all detectors, therefore you only need to know how to enumerate over them. Passing a predicate to aid a search might be a useful option.